# INTRODUCTION
This project analyzes a dataset from the Google Play Store containing information on thousands of mobile applications. The main objective is to perform exploratory data analysis (EDA) by cleaning and transforming the dataset, engineering relevant features, and creating visualizations to uncover valuable insights into customer behavior and app preferences.

As the project is intended for showcasing and practicing purposes, some steps (mainly in the FE section) are not strictly necessary for the completion of the tasks or the final analysis.  
These intentional additions are meant to show familiraity with different data-preparation techniques.  
Additonally, while data cleaning and the feature engineering are often to be intertwined and occur simultenously in a real workflow, separating them into distinct sections was a deliberate choice to improve structure and clarity

**Data source**
The dataset was provided by udemy

## Tasks
a) Perform a general univariate analysis  
b) What are the most common prices for apps  
c) Hat are the most present apps in google playstore   
d) Which category has the largest number of installs  
e) What are th top most installed apps in each popular category

## Scope of the project
This study is part of a broader portfolio project designed to showcase my data analysis skills, including:
- Statistical reasoning
- Proficiency with core Python libraries (NumPy, Pandas, Matplotlib, Seaborn)
- Automation principles
- Overall analytical approach to working with structured data

The study is structured as follows:
1. Data Cleaning
2. Feature engineering
3. Exploratory data analysis

### Data cleaning
This section focuses on preparing the dataset for further analysis by identifying and handling duplicate records and missing values. It also includes correcting formatting issues and addressing any inconsistencies in the data.

### Feature enginerring
In this section, the dataset is transformed to include new or restructured variables that improve its suitability for analysis or modeling. This is a relatively small part of the project, as machine learning is not the final goal and most features are already formatted in a modeling-friendly way from the source.

### Exploratory data analysis
This section contains the core analytical work, broken into two complementary steps:

- Code-based exploration – Direct investigation using code and printed outputs, providing a quick, tabular overview of key findings.
- Visual exploration – Graphical representations of the same tasks to enhance clarity and allow for faster, more intuitive interpretation of the results.

By combining both approaches, the analysis aims to uncover meaningful insights into user behavior and app characteristics on the Google Play Store. 


# DATA CLEANING

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import seaborn as sns
import warnings

warnings.filterwarnings("ignore")


In [5]:
import matplotlib as mpl
mpl.rcParams.update(mpl.rcParamsDefault)

In [6]:
df = pd.read_csv('https://raw.githubusercontent.com/krishnaik06/playstore-Dataset/main/googleplaystore.csv')

## 1. Understanding data structure

### 1.1 Getting familiar with the dataset

In [11]:
# Displaying the dataset 
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [14]:
# Initial number of max rows (records) and column (features)
df.shape

(10841, 13)

In [16]:
# Retrieving general information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


### Insights -Getting Familiar with dataset
The output of .info() on the full dataset shows that several features contain missing values.
It also reveals that Rating is the only column stored as a numeric type, while other columns currently stored as strings should be converted to more appropriate data types.

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
### 1.2 Finding unique values
This step aims to identify the unique values for each feature in the dataset.

Using .unique() is helpful for understanding the data type of a feature and gaining a preliminary view of how its records are structured and formatted.
However, when features have a large number of unique values, the output becomes difficult to interpret at a glance. To keep the results readable and concised, the following code excludes features with more than 50 unique entries.

That said, it's generally more effective to inspect a single feature individually using .unique() when a specific need arises.

In [20]:
df.nunique()

App               9660
Category            34
Rating              40
Reviews           6002
Size               462
Installs            22
Type                 3
Price               93
Content Rating       6
Genres             120
Last Updated      1378
Current Ver       2832
Android Ver         33
dtype: int64

In [22]:
for col in df:
        unique_count = df[col].nunique()
        
        if df[col].nunique() < 50:
            unique_value = []
            unique_value.append(df[col].unique())
            print('----------------------------------------')
            print(f"\n\033[91mThe unique values for {col} are:\033[0m\n\n {unique_value}")

----------------------------------------

The unique values for Category are:

 [array(['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY',
       'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION',
       'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE',
       'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME',
       'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL',
       'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL',
       'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER',
       'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION',
       '1.9'], dtype=object)]
----------------------------------------

The unique values for Rating are:

 [array([ 4.1,  3.9,  4.7,  4.5,  4.3,  4.4,  3.8,  4.2,  4.6,  3.2,  4. ,
        nan,  4.8,  4.9,  3.6,  3.7,  3.3,  3.4,  3.5,  3.1,  5. ,  2.6,
        3. ,  1.9,  2.5,  2.8,  2.7,  1. ,  2.9,  2.3,  2.2,  1.7,  2. ,
        1.8,  2.4,  1.6,  2.1,  1.4,  1.5,  

## Insights - Data structure


1. **App**
- Current Type: Categorical (object)
- Description: Name of the app.
- Unique values: 9,660; No missing values.
- No conversion needed

2. **Category**
- Current Type: Categorical (object)
- Description: Broad app category (e.g., ART_AND_DESIGN).
- Unique values: 36; No missing values.
- No conversion needed

3. **Rating**
- Current Type: Numerical (float64)
- Description: Average user rating (0.0 to 5.0).
- Unique values: 40; multiple missing values
- Already numeric – may require imputation for missing values.

5. **Reviews**
- Current Type: Categorical (object)
- Description: Total user reviews per app, stored as string.
- Unique values: 6,002; No missing values
- To be converted to numerical (integer)

6. **Size**
- Cuttent Type: Categorical (object)
- Description: App size in MB or KB (e.g., “19M”, “14k”, “Varies with device”).
- Unique values: 462; No missing values
- To be converted to numerical (float in MB; handle special values like "Varies with device")

7. **Installs**
- Current Type: Categorical (object)
- Description: App installs/downloads (e.g., “1,000+”).
- Unique values: 22; No missing values
- To be converted to integer (remove “+” and “,”)

8. **Type**
- Current Type: Categorical (object)
- Description: Free or Paid app.
- Unique values: 3; 1 missing value
- Can stay categorical – optionally encoded.

9. **Price**
- Curretn Type: Categorical (object)
- Description: App price (e.g., “$0.00”)
- Unique values: 93; No missing values
- To be converted to float (remove dollar sign)

10. **Content Rating**
- Current Type: Categorical (object)
- Description: Minimum age group (e.g., Everyone, Teen)
- Unique values: 6; 1 missing value
- Can stay categorical – optionally encoded.

11. **Genres**
- Current Type: Categorical (object)
- Description: App subcategory/genre (some are compound, e.g., “Art & Design;Pretend Play”)
- Unique values: 120; No missing values
- Can stay categorical – optionally encoded.

12. **Last Updated**
- Current Type: Categorical (object)
- Description: Last update date (e.g., “January 7, 2018”)
- Unique values: 1,378; No missing values
- To be converted to datetime

13. **Current Ver**
- Current Type: Categorical (object)
- Description: App version (e.g., “1.0.0”)
- Unique values: 2,832; few missing values
- Can remain as is – though version parsing is possible.

14. **Android Ver**
- Type: Categorical (object)
- Description: Minimum Android OS required (e.g., “4.0.3 and up”)
- Unique values: 33; few 3 missing values
- Can be cleaned and optionally converted to float or category or encoded

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

##### It's good practice to make a copy of the original dataset to work on

In [27]:
df_copy = df.copy()

## 2. Identyfing and handling duplicates (preliminary inspection)
This preliminary inspection is based on two main assumptions:
1. Identical duplicate rows are irrelevant
2. The feature App is a convenient primary key
***

Duplicated rows do not carry additional information in this dataset—each row represents a unique record, and if all metadata for an app is identical, it is considered a redundant repetition rather than a distinct instance of different behavior.  
Therefore, **it is safe to delete such redundant repetitions without losing information about the dataset!**
***
The App feature should act as a primary key, meaning each row should correspond to a unique app, with all associated metadata referring to that single entity.  
Therefore, the first step is to determine whether all duplicated rows share the same app name.

- If every duplicate row corresponds to a duplicated App, then removing exact duplicates would leave App as a valid primary key.
- However, if the number of duplicated rows differs from the number of duplicated App values, this suggests the presence of distinct rows sharing the same app name but with different metadata—indicating inconsistent or conflicting records.  

In the second case, **further analysis is required to understand the discrepancies among repeated app entries and decide on an appropriate strategy for handling them**.

#### 2.1 Identical duplicates

In [31]:
# Number of identical duplicate rows including the first instance
tot_dup = df_copy.duplicated(keep = False).sum()

In [33]:
# Number of identical duplicate rows excluding the first instance 
dup = df_copy.duplicated().sum()

In [35]:
max_rows = df_copy.shape[0]

print(f"\n\033[91mNumber of identical duplicate rows - including first instance:\033[0m\n {tot_dup}")
print(f"\n\033[91mNumber of identical duplicate row - excluding first instance:\033[0m\n {dup}")
print(f"\n\033[91mTotal number of rows per columns:\033[0m\n {max_rows}")


Number of identical duplicate rows - including first instance:
 893

Number of identical duplicate row - excluding first instance:
 483

Total number of rows per columns:
 10841


In [37]:
# Percentage of identical duplicate rows (excluing first instance
percentage_dup = (dup/max_rows)*100
print(f"\n\033[91mPercentage of duplicate rows respect to the total dataset - excluding first instance:\033[0m\n {round(percentage_dup, 2)}")


Percentage of duplicate rows respect to the total dataset - excluding first instance:
 4.46


***
#### 2.2 Duplicate apps with different metadata

In [261]:
# Number of duplicate apps including the first instance
tot_dup_all = df_copy['App'].duplicated(keep = False).sum()

In [263]:
# Number of duplicate apps excluding the first instance
tot_dup = df_copy['App'].duplicated().sum()

In [265]:
# Number of unique app names
print(f"\n\033[91mNumber of duplicated apps with different metadata - including first instance:\033[0m\n {tot_dup_all}")
print(f"\n\033[91mNumber of duplicated apps with different metadata - excluding first instance:\033[0m\n {tot_dup}")
print(f"\n\033[91mNumber of unique apps:\033[0m\n {df_copy['App'].nunique()}")


Number of duplicated apps with different metadata - including first instance:
 1979

Number of duplicated apps with different metadata - excluding first instance:
 1181

Number of unique apps:
 9660


In [267]:
# Double check for app that have the same name and are included in the general duplicates
df_copy[df_copy['App'].duplicated(keep=False) & df_copy.duplicated(keep=False)].shape #redundant

(893, 13)

The previous check tells that all the duplicated rows have also duplicated apps, as it was forseable.

#### 2.3 Dropping identical duplicated rows, expect first instance

In [271]:
# Dropping all the duplicate rows except the first instance
df_copy.drop_duplicates(keep = 'first', inplace = True)

In [273]:
# Checking if the drop_duplicates was performed succesfully
df_copy.duplicated().sum()

0

In [275]:
print(f"\n\033[91mNumber of remaining rows after identical duplicates drop:\033[0m\n{df_copy.shape[0]}")


Number of remaining rows after identical duplicates drop:
10358


In [277]:
dup_app = df_copy['App'].duplicated().sum()
dup_app_all = df_copy['App'].duplicated(keep = False).sum()

print(f"\n\033[91mTotal records with duplicated app names after drop - including first instance:\033[0m\n {dup_app_all}")
print(f"\n\033[91mNumber of records with duplicated app names after drop - excluding first instance:\033[0m\n {dup_app}")


Total records with duplicated app names after drop - including first instance:
 1221

Number of records with duplicated app names after drop - excluding first instance:
 698


In [279]:
print("\n\033[91mCount of duplicate apps with different metadata:\033[0m\n")
df_copy[df_copy['App'].duplicated(keep = False)]['App'].value_counts()


Count of duplicate apps with different metadata:



App
ROBLOX                                           9
8 Ball Pool                                      7
Helix Jump                                       6
Zombie Catchers                                  6
Bubble Shooter                                   6
                                                ..
Nike Training Club - Workouts & Fitness Plans    2
Weight Loss Running by Verv                      2
Run with Map My Run                              2
Endomondo - Running & Walking                    2
Maps & GPS Navigation — OsmAnd                   2
Name: count, Length: 523, dtype: int64

## Insigths - duplicates

The number of duplicates identified at the dataset level does not fully match the count for the App feature, which indicates that some applications share the same name but differ in other attributes.

Although duplicate rows can be safely removed based on the considerations outlined at the beginning of this chapter, it remains important to further investigate the App feature to understand the source of these repetitions.

In any case, the proportion of duplicates relative to the total dataset (4.4%) is not high enough to raise concerns about the data collection process, nor would their removal represent a significant loss of information for this type of study.

The additional analysis will be included at the start of the EDA section.

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
## 3. Identifying and handling missing values (preliminary inspection)
As informally indicated by the info in the first chapter, there are multiple features presenting missing values in different amounts.  
This section will focus on identifying such records and draw some initial insights and conclusions.
The proper analysis will be also conducted in the EDA chapter.

In [283]:
import io

datasets = [("df", df), ("df_copy", df_copy)]

for name, dtset in datasets:
    buffer = io.StringIO()
    dtset.info(buf=buffer)   # send output to buffer instead of printing
    info_str = buffer.getvalue()
    print('----------------------------------------')

    print(f"\033[91mDataset info for {name}:\033[0m\n")
    
    print(info_str)


----------------------------------------
Dataset info for df:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB

----------------------------------------
Dataset info for df_copy:

<class 'pandas.core.frame.DataF

In [285]:
for name, dtset in datasets:
    print('----------------------------------------')
    print(f"\033[91mMissing values for {name}:\033[0m\n")
    print(dtset.isnull().sum())


----------------------------------------
Missing values for df:

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64
----------------------------------------
Missing values for df_copy:

App                  0
Category             0
Rating            1465
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64


In [287]:
rating_null = df_copy['Rating'].isna().sum()

print(f"\033[91mNumber of null values in Rating:\033[0m\n {rating_null}\n\n\033[91mNumber of total unique rows in the dataset:\033[0m\n {df_copy.shape[0]}")
print(f"\033[91m\nPercentage of missing values in Rating compared to the total dataset:\033[0m\n {(rating_null/df_copy.shape[0])*100}")

Number of null values in Rating:
 1465

Number of total unique rows in the dataset:
 10358

Percentage of missing values in Rating compared to the total dataset:
 14.143657076655725


In [41]:
col_null = ['Type', 'Content Rating', 'Current Ver', 'Android Ver']

for col in col_null:
    num_null = df_copy[col].isnull().sum()
    print('\n----------------------------------------')
    print(f"\033[91m\nNumber of missing values in {col}:\033[0m\n {num_null}")
    print(f"\033[91m\nPercentage of missing values compared to the entire dataset for {col}:\033[0m\n {round((num_null/df_copy.shape[0])*100, 3)}")


----------------------------------------

Number of missing values in Type:
 1

Percentage of missing values compared to the entire dataset for Type:
 0.009

----------------------------------------

Number of missing values in Content Rating:
 1

Percentage of missing values compared to the entire dataset for Content Rating:
 0.009

----------------------------------------

Number of missing values in Current Ver:
 8

Percentage of missing values compared to the entire dataset for Current Ver:
 0.074

----------------------------------------

Number of missing values in Android Ver:
 3

Percentage of missing values compared to the entire dataset for Android Ver:
 0.028


## Insights - missing values
As briefly outlined in the first chapter trought the .info() on the entire dataset, several features contain missing values in varying proportions.  

Especially Rating shows a concerning ratio of missing/total values amounting to 14.1% despite the duplicates dropping.
For the remaining features presenting missing values, the percentage compared to the entire dataset is neglible as it's under 1% of the total data: the rows containing these missing values can be confidently dropped.

In [390]:
# Dropping the rows with for features containing small proportion of NaN
df_copy.dropna(axis = 0, subset = ['Type', 'Content Rating', 'Current Ver', 'Android Ver'], inplace = True)

In [392]:
# Checking if the dropping process worked
print('\033[91m\nCheck on number of missing value per feature after drop:\033[0m')
df_copy[['Type', 'Content Rating', 'Current Ver', 'Android Ver']].isna().sum()


Check on number of missing value per feature after drop:


Type              0
Content Rating    0
Current Ver       0
Android Ver       0
dtype: int64

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
## 4. Converting data types

### 4.1 Converting 'Reviews'

In [295]:
# Checking how many numeric and no numeric strings are in Reviews 
df_copy['Reviews'].str.isnumeric().value_counts().to_frame()

count
Reviews       
True     10357
False        1

In [297]:
# Converting Reviews to int
df_copy['Reviews'] = df_copy['Reviews'].astype(int)

ValueError: invalid literal for int() with base 10: '3.0M'

In [ ]:
# Retrieving the only non-numeric record in Reviews
df[~df['Reviews'].str.isnumeric()]

### Insights - converting Reviews

Here there's only one row not containing purely numbers: in fact, if we take a look at df['Reviews'].info() there are 10841 rows while, as we can see from df['Reviews'].str.isnumeric().sum(), there are 10840 rows containing only numbers. 
This leaves us with one row with a string containing at least something that is not a number

We can Either:
- drop the row
- convert 3M to number

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
#### 4.1.1 Handling the unusual record in Reviews

In [308]:
# Converting the unusual record to int 
df_copy.loc[df['Reviews']=='3.0M', 'Reviews'] = 30000000

#We can use 2 more approaches:
#df_copy.iloc[10472, df_copy.columns.get_loc('Reviews')] = 30000000

#or

#df_copy.at[10472, 'Reviews'] = 3000000

In [310]:
# Checking if the row ws correctly removed
df_copy[df_copy['Reviews']=='3.0']

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver]
Index: []

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

#### 4.1.2 Final Conversion

In [314]:
# Converting Reviews to int
df_copy['Reviews'] = df_copy['Reviews'].astype(int)

In [316]:
print('\033[91m\nChecking if conversion worked properly\033[0m\n')
df_copy['Reviews'].info()


Checking if conversion worked properly

<class 'pandas.core.series.Series'>
Index: 10358 entries, 0 to 10840
Series name: Reviews
Non-Null Count  Dtype
--------------  -----
10358 non-null  int64
dtypes: int64(1)
memory usage: 161.8 KB


🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
### 4.2 Converting 'Size'

In [319]:
print('\033[91m\nRetrievening unique values for Size\033[0m\n')
df_copy['Size'].unique()


Retrievening unique values for Size



array(['19M', '14M', '8.7M', '25M', '2.8M', '5.6M', '29M', '33M', '3.1M',
       '28M', '12M', '20M', '21M', '37M', '2.7M', '5.5M', '17M', '39M',
       '31M', '4.2M', '7.0M', '23M', '6.0M', '6.1M', '4.6M', '9.2M',
       '5.2M', '11M', '24M', 'Varies with device', '9.4M', '15M', '10M',
       '1.2M', '26M', '8.0M', '7.9M', '56M', '57M', '35M', '54M', '201k',
       '3.6M', '5.7M', '8.6M', '2.4M', '27M', '2.5M', '16M', '3.4M',
       '8.9M', '3.9M', '2.9M', '38M', '32M', '5.4M', '18M', '1.1M',
       '2.2M', '4.5M', '9.8M', '52M', '9.0M', '6.7M', '30M', '2.6M',
       '7.1M', '3.7M', '22M', '7.4M', '6.4M', '3.2M', '8.2M', '9.9M',
       '4.9M', '9.5M', '5.0M', '5.9M', '13M', '73M', '6.8M', '3.5M',
       '4.0M', '2.3M', '7.2M', '2.1M', '42M', '7.3M', '9.1M', '55M',
       '23k', '6.5M', '1.5M', '7.5M', '51M', '41M', '48M', '8.5M', '46M',
       '8.3M', '4.3M', '4.7M', '3.3M', '40M', '7.8M', '8.8M', '6.6M',
       '5.1M', '61M', '66M', '79k', '8.4M', '118k', '44M', '695k', '1.6M',
     

In [321]:
df_copy[['Size']]

Size
0                     19M
1                     14M
2                    8.7M
3                     25M
4                    2.8M
...                   ...
10836                 53M
10837                3.6M
10838                9.5M
10839  Varies with device
10840                 19M

[10358 rows x 1 columns]

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

In [324]:
# Converting the M and k to number and 'Varies with device' to NaN
df_copy['Size'] = df_copy['Size'].str.replace('M','000000')
df_copy['Size'] = df_copy['Size'].str.replace('k','000')
df_copy['Size'] = df_copy['Size'].replace('Varies with device', np.nan)#.astype(float)
df_copy['Size'] = df_copy['Size'].astype(float)
df_copy.head(2)

ValueError: could not convert string to float: '1,000+'

**NOTE**
Another ValueError has been raised, higlighting the presence of a unexpected format for one record in Size

**The error has been intentionally left as a prove of the process.**

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
##### 4.2.1 Handling the unusual record in 'Size'

In [328]:
df_copy[df_copy['Size'] == '1,000+']

App Category  Rating   Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0  30000000   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

**NOTE**:
this is the same row from before earlier when the issue of Reviews having a record store as '3.0M'. 
Given the fact that imptutating 1000+ is impossible and it's only row, it appears reasonable to just drop the row at once.

In [331]:
df_copy.drop(10472, inplace = True)

In [333]:
df_copy[df_copy['Size'] == '1,000+']

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver]
Index: []

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
##### 4.2.2 Final conversion

In [336]:
#We can unify all the M (=milion=000000) and K (=thousand=000)
df_copy['Size'] = df_copy['Size'].str.replace('M','000000')
df_copy['Size'] = df_copy['Size'].str.replace('k','000')
df_copy['Size'] = df_copy['Size'].replace('Varies with device', np.nan)
df_copy['Size'] = df_copy['Size'].astype(float)
df_copy.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews        Size  Installs  Type Price Content Rating  \
0      159  19000000.0   10,000+  Free     0       Everyone   
1      967  14000000.0  500,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up

In [338]:
print('\033[91m\nChecking if conversion worked properly\033[0m\n')
df_copy['Size'].info()


Checking if conversion worked properly

<class 'pandas.core.series.Series'>
Index: 10357 entries, 0 to 10840
Series name: Size
Non-Null Count  Dtype  
--------------  -----  
8831 non-null   float64
dtypes: float64(1)
memory usage: 161.8 KB


🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
### 4.3 Converting 2 features at the same time with loops: 'Installs' and 'Price'

In [341]:
columns = ['Price', 'Installs']

for col in columns:
    print(f"\033[91m\nRetrievening unique values for {col}:\033[0m\n{df_copy[col].unique()}")


Retrievening unique values for Price:
['0' '$4.99' '$3.99' '$6.99' '$1.49' '$2.99' '$7.99' '$5.99' '$3.49'
 '$1.99' '$9.99' '$7.49' '$0.99' '$9.00' '$5.49' '$10.00' '$24.99'
 '$11.99' '$79.99' '$16.99' '$14.99' '$1.00' '$29.99' '$12.99' '$2.49'
 '$10.99' '$1.50' '$19.99' '$15.99' '$33.99' '$74.99' '$39.99' '$3.95'
 '$4.49' '$1.70' '$8.99' '$2.00' '$3.88' '$25.99' '$399.99' '$17.99'
 '$400.00' '$3.02' '$1.76' '$4.84' '$4.77' '$1.61' '$2.50' '$1.59' '$6.49'
 '$1.29' '$5.00' '$13.99' '$299.99' '$379.99' '$37.99' '$18.99' '$389.99'
 '$19.90' '$8.49' '$1.75' '$14.00' '$4.85' '$46.99' '$109.99' '$154.99'
 '$3.08' '$2.59' '$4.80' '$1.96' '$19.40' '$3.90' '$4.59' '$15.46' '$3.04'
 '$4.29' '$2.60' '$3.28' '$4.60' '$28.99' '$2.95' '$2.90' '$1.97'
 '$200.00' '$89.99' '$2.56' '$30.99' '$3.61' '$394.99' '$1.26' '$1.20'
 '$1.04']

Retrievening unique values for Installs:
['10,000+' '500,000+' '5,000,000+' '50,000,000+' '100,000+' '50,000+'
 '1,000,000+' '10,000,000+' '5,000+' '100,000,000+' '1,000,

In [343]:
# Replacing unwanted symbols before convertion
char_to_remove = ['$','+',',']

for item in char_to_remove:
    for col in columns:
        df_copy[col] = df_copy[col].str.replace(item,'')

for col in columns:
    unique = df_copy[col].unique()
    print(f"\033[91m\nRetrievening unique values for {col} after replacing:\033[0m\n\n{unique}")



Retrievening unique values for Price after replacing:

['0' '4.99' '3.99' '6.99' '1.49' '2.99' '7.99' '5.99' '3.49' '1.99' '9.99'
 '7.49' '0.99' '9.00' '5.49' '10.00' '24.99' '11.99' '79.99' '16.99'
 '14.99' '1.00' '29.99' '12.99' '2.49' '10.99' '1.50' '19.99' '15.99'
 '33.99' '74.99' '39.99' '3.95' '4.49' '1.70' '8.99' '2.00' '3.88' '25.99'
 '399.99' '17.99' '400.00' '3.02' '1.76' '4.84' '4.77' '1.61' '2.50'
 '1.59' '6.49' '1.29' '5.00' '13.99' '299.99' '379.99' '37.99' '18.99'
 '389.99' '19.90' '8.49' '1.75' '14.00' '4.85' '46.99' '109.99' '154.99'
 '3.08' '2.59' '4.80' '1.96' '19.40' '3.90' '4.59' '15.46' '3.04' '4.29'
 '2.60' '3.28' '4.60' '28.99' '2.95' '2.90' '1.97' '200.00' '89.99' '2.56'
 '30.99' '3.61' '394.99' '1.26' '1.20' '1.04']

Retrievening unique values for Installs after replacing:

['10000' '500000' '5000000' '50000000' '100000' '50000' '1000000'
 '10000000' '5000' '100000000' '1000000000' '1000' '500000000' '50' '100'
 '500' '10' '1' '5' '0']


In [345]:
# Converting both features first to float
for col in columns:
    df_copy[col]=df_copy[col].astype(float)

In [347]:
print('\033[91m\nCheck if convertion to float worked\033[0m\n')
df_copy.info()


Check if convertion to float worked

<class 'pandas.core.frame.DataFrame'>
Index: 10357 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10357 non-null  object 
 1   Category        10357 non-null  object 
 2   Rating          8892 non-null   float64
 3   Reviews         10357 non-null  int64  
 4   Size            8831 non-null   float64
 5   Installs        10357 non-null  float64
 6   Type            10356 non-null  object 
 7   Price           10357 non-null  float64
 8   Content Rating  10357 non-null  object 
 9   Genres          10357 non-null  object 
 10  Last Updated    10357 non-null  object 
 11  Current Ver     10349 non-null  object 
 12  Android Ver     10355 non-null  object 
dtypes: float64(4), int64(1), object(8)
memory usage: 1.1+ MB


In [349]:
# Converting Installs to int
df_copy['Installs'] = df_copy['Installs'].astype(int)
print('\033[91m\nConverting Installs to int and check:\033[0m\n')
df_copy['Installs'].info()


Converting Installs to int and check:

<class 'pandas.core.series.Series'>
Index: 10357 entries, 0 to 10840
Series name: Installs
Non-Null Count  Dtype
--------------  -----
10357 non-null  int64
dtypes: int64(1)
memory usage: 161.8 KB


In [351]:
df_copy.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews        Size  Installs  Type  Price Content Rating  \
0      159  19000000.0     10000  Free    0.0       Everyone   
1      967  14000000.0    500000  Free    0.0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
### 4.4 Convertine 'Last Update' to timestamp

In [354]:
print(f"\033[91m\nRetrievening unique values for Last Updated:\033[0m\n{df_copy['Last Updated'].unique()}")


Retrievening unique values for Last Updated:
['January 7, 2018' 'January 15, 2018' 'August 1, 2018' ...
 'January 20, 2014' 'February 16, 2014' 'March 23, 2014']


In [356]:
# Converting Last Update to dtetime
df_copy['Last Updated'] = pd.to_datetime(df_copy['Last Updated'])
print('\033[91m\nCheck if convertion to datetime worked:\033[0m\n')
df_copy.info()


Check if convertion to datetime worked:

<class 'pandas.core.frame.DataFrame'>
Index: 10357 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   App             10357 non-null  object        
 1   Category        10357 non-null  object        
 2   Rating          8892 non-null   float64       
 3   Reviews         10357 non-null  int64         
 4   Size            8831 non-null   float64       
 5   Installs        10357 non-null  int64         
 6   Type            10356 non-null  object        
 7   Price           10357 non-null  float64       
 8   Content Rating  10357 non-null  object        
 9   Genres          10357 non-null  object        
 10  Last Updated    10357 non-null  datetime64[ns]
 11  Current Ver     10349 non-null  object        
 12  Android Ver     10355 non-null  object        
dtypes: datetime64[ns](1), float64(3), int64(2), object(7)
memory usage: 1

In [358]:
df_copy.head(2)

App        Category  Rating  \
0  Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                             Coloring book moana  ART_AND_DESIGN     3.9   

   Reviews        Size  Installs  Type  Price Content Rating  \
0      159  19000000.0     10000  Free    0.0       Everyone   
1      967  14000000.0    500000  Free    0.0       Everyone   

                      Genres Last Updated Current Ver   Android Ver  
0               Art & Design   2018-01-07       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play   2018-01-15       2.0.0  4.0.3 and up

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
# FEATURE ENGINEERING
This section will be relatively short as most Features are already prepared for machine learning modeling

## 1. Extracting info

In [362]:
# Extracting day, month and year from Last Updated
df_copy['Date'] = df_copy['Last Updated'].dt.day
df_copy['Month'] = df_copy['Last Updated'].dt.month
df_copy['Year'] = df_copy['Last Updated'].dt.year

In [364]:
print('\033[91m\nCheck if extraction worked:\033[0m\n')
df_copy.info()


Check if extraction worked:

<class 'pandas.core.frame.DataFrame'>
Index: 10357 entries, 0 to 10840
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   App             10357 non-null  object        
 1   Category        10357 non-null  object        
 2   Rating          8892 non-null   float64       
 3   Reviews         10357 non-null  int64         
 4   Size            8831 non-null   float64       
 5   Installs        10357 non-null  int64         
 6   Type            10356 non-null  object        
 7   Price           10357 non-null  float64       
 8   Content Rating  10357 non-null  object        
 9   Genres          10357 non-null  object        
 10  Last Updated    10357 non-null  datetime64[ns]
 11  Current Ver     10349 non-null  object        
 12  Android Ver     10355 non-null  object        
 13  Date            10357 non-null  int32         
 14  Month           10357 non-nul

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
## 2. Creating new features

### 2.1 Creating a new Feature for Current Ver

In [368]:
# Converting Current Ver and filling NaN
df_copy['Current Ver']=df_copy['Current Ver'].astype(str).fillna('')

In [370]:
# Replacing sy
df_copy['Current Ver'] = df_copy['Current Ver'].str.replace('.','', regex = False)
df_copy[~df_copy['Current Ver'].str.isnumeric()]

App             Category  \
3                              Sketch - Draw & Paint       ART_AND_DESIGN   
15                   Learn To Draw Kawaii Characters       ART_AND_DESIGN   
37                                Floor Plan Creator       ART_AND_DESIGN   
42                        Textgram - write on photos       ART_AND_DESIGN   
52                     Used Cars and Trucks for Sale    AUTO_AND_VEHICLES   
...                                              ...                  ...   
10725                                      Posta App  MAPS_AND_NAVIGATION   
10765                Chat For Strangers - Video Chat               SOCIAL   
10826      Frim: get new friends on local chat rooms               SOCIAL   
10839                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating  Reviews        Size  Installs  Type  Price Content Rating  \
3         4.5   215644  25000000.0  50000000  Free    0.0           Teen   
15        3.2       55         2.7      5000  Free    0.0       Everyone   
37        4.1    36639         NaN   5000000  Free    0.0       Everyone   
42        4.4   295221         NaN  10000000  Free    0.0       Everyone   
52        4.6    17057         NaN   1000000  Free    0.0       Everyone   
...       ...      ...         ...       ...   ...    ...            ...   
10725     3.6        8         NaN      1000  Free    0.0       Everyone   
10765     3.4      622         NaN    100000  Free    0.0     Mature 17+   
10826     4.0    88486         NaN   5000000  Free    0.0     Mature 17+   
10839     4.5      114         NaN      1000  Free    0.0     Mature 17+   
10840     4.5   398307  19000000.0  10000000  Free    0.0       Everyone   

                  Genres Last Updated         Current Ver         Android Ver  \
3           Art & Design   2018-06-08  Varies with device          4.2 and up   
15          Art & Design   2018-06-06                 nan          4.2 and up   
37          Art & Design   2018-07-14  Varies with device        2.3.3 and up   
42          Art & Design   2018-07-30  Varies with device  Varies with device   
52       Auto & Vehicles   2018-07-30  Varies with device  Varies with device   
...                  ...          ...                 ...                 ...   
10725  Maps & Navigation   2017-09-27  Varies with device          4.4 and up   
10765             Social   2018-05-23  Varies with device  Varies with device   
10826             Social   2018-03-23  Varies with device  Varies with device   
10839  Books & Reference   2015-01-19  Varies with device  Varies with device   
10840          Lifestyle   2018-07-25  Varies with device  Varies with device   

       Date  Month  Year  
3         8      6  2018  
15        6      6  2018  
37       14      7  2018  
42       30      7  2018  
52       30      7  2018  
...     ...    ...   ...  
10725    27      9  2017  
10765    23      5  2018  
10826    23      3  2018  
10839    19      1  2015  
10840    25      7  2018  

[1705 rows x 16 columns]

In [372]:
df_copy['Current Ver'].unique()

array(['100', '200', '124', ..., '10612928', '034', '201480'],
      dtype=object)

In [374]:
#Creating new Features from Current Ver
df_copy['Major'] = df_copy['Current Ver'].str.split('.', n=1).str[0]
df_copy['Minor'] = df_copy['Current Ver'].str.split('.', n=2).str[1]
df_copy['Patch'] = df_copy['Current Ver'].str.split('.', n=3).str[2]

In [376]:
df_copy.head(10)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   
5                         Paper flowers instructions  ART_AND_DESIGN     4.4   
6            Smoke Effect Photo Maker - Smoke Editor  ART_AND_DESIGN     3.8   
7                                   Infinite Painter  ART_AND_DESIGN     4.1   
8                               Garden Coloring Book  ART_AND_DESIGN     4.4   
9                      Kids Paint Free - Drawing Fun  ART_AND_DESIGN     4.7   

   Reviews        Size  Installs  Type  Price Content Rating  \
0      159  19000000.0     10000  Free    0.0       Everyone   
1      967  14000000.0    500000  Free    0.0       Everyone   
2    87510         8.7   5000000  Free    0.0       Everyone   
3   215644  25000000.0  50000000  Free    0.0           Teen   
4      967         2.8    100000  Free    0.0       Everyone   
5      167         5.6     50000  Free    0.0       Everyone   
6      178  19000000.0     50000  Free    0.0       Everyone   
7    36815  29000000.0   1000000  Free    0.0       Everyone   
8    13791  33000000.0   1000000  Free    0.0       Everyone   
9      121         3.1     10000  Free    0.0       Everyone   

                      Genres Last Updated         Current Ver   Android Ver  \
0               Art & Design   2018-01-07                 100  4.0.3 and up   
1  Art & Design;Pretend Play   2018-01-15                 200  4.0.3 and up   
2               Art & Design   2018-08-01                 124  4.0.3 and up   
3               Art & Design   2018-06-08  Varies with device    4.2 and up   
4    Art & Design;Creativity   2018-06-20                  11    4.4 and up   
5               Art & Design   2017-03-26                  10    2.3 and up   
6               Art & Design   2018-04-26                  11  4.0.3 and up   
7               Art & Design   2018-06-14               61611    4.2 and up   
8               Art & Design   2017-09-20                 292    3.0 and up   
9    Art & Design;Creativity   2018-07-03                  28  4.0.3 and up   

   Date  Month  Year               Major  Minor  Patch  
0     7      1  2018                 100    NaN    NaN  
1    15      1  2018                 200    NaN    NaN  
2     1      8  2018                 124    NaN    NaN  
3     8      6  2018  Varies with device    NaN    NaN  
4    20      6  2018                  11    NaN    NaN  
5    26      3  2017                  10    NaN    NaN  
6    26      4  2018                  11    NaN    NaN  
7    14      6  2018               61611    NaN    NaN  
8    20      9  2017                 292    NaN    NaN  
9     3      7  2018                  28    NaN    NaN

### 2.2 Encoding

In [379]:
from sklearn.preprocessing import OneHotEncoder

In [381]:
encoder = OneHotEncoder()

In [383]:
encoded = encoder.fit_transform(df_copy[['Content Rating']]).toarray()
encoded

array([[0., 1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0.],
       ...,
       [0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 1., 0., 0., 0., 0.]])

In [385]:
encoded_df= pd.DataFrame(encoded, columns = encoder.get_feature_names_out())
encoded_df.head(5)

Content Rating_Adults only 18+  Content Rating_Everyone  \
0                             0.0                      1.0   
1                             0.0                      1.0   
2                             0.0                      1.0   
3                             0.0                      0.0   
4                             0.0                      1.0   

   Content Rating_Everyone 10+  Content Rating_Mature 17+  \
0                          0.0                        0.0   
1                          0.0                        0.0   
2                          0.0                        0.0   
3                          0.0                        0.0   
4                          0.0                        0.0   

   Content Rating_Teen  Content Rating_Unrated  
0                  0.0                     0.0  
1                  0.0                     0.0  
2                  0.0                     0.0  
3                  1.0                     0.0  
4                  0.0                     0.0

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

In [388]:
df_copy.to_parquet('df_copy.parquet')